### DISCUSSION TITLE
v3 fails to reproduce any matches where v2 did

### DISCUSSION THREAD

**Discussion by akraemer007** [moj-analytical-services/splink#855] [Q&A] [answered]
### What happens?

I am updating my spark splink pipeline from v2 - > v3. Using the same blocking rules, comparisons, settings, and underlying data, I am getting the warning below for all of my columns when running `linker.estimate_u_using_random_sampling(target_rows=1e15)`:

`INFO:splink.m_u_records_to_parameters:u probability not trained for xxxxxx- Exact match (comparison vector value: 1). This usually means the comparison level was never observed in the training data.`

When I run my predict (after using expectation_maximization which similarly has no matches), I get results saying the highest probability of mach is nearly zero, whereas I was getting a nice distribution and many high probability matches in v2.

I'm wondering if this has anything to do with there being no blocking rules applied when estimating u as shown below -- rather than the blocking rules i used.

![image](https://user-images.githubusercontent.com/12884043/196733289-2389e70a-71fe-477a-b618-9d83c22d70a9.png)

Any guidance would be greatly appreciated!

### To Reproduce

Not sure how to reproduce, since it works fine with the demo data

### OS:

DataBricks

### Splink version:

3.1.4

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by RobinL** [comment, marked as answer]
We would need to see the full settings config for the v2 and v3 models to help with this, and ideally the full Splink script

Blocking rules are not applied for training u values - the rationale for this is described in a bit more detail at the following links:

https://towardsdatascience.com/the-intuition-behind-the-use-of-expectation-maximisation-to-train-record-linkage-models-5d7ac9f019ca
https://moj-analytical-services.github.io/splink/linker.html#splink.linker.Linker.estimate_u_using_random_sampling

I'm going to migrate this to a discussion for now - can always make it an issue again if we find a bug

**Reply by akraemer007** [reply]
Well, it turns out the only bug is me. In copying all of my Splink code to share, I found my typo. It looks like I imported the spark comparison library and the for the duckdb comparion level library.

I was truly losing my mind. Who would have guessed it was the imports! Thanks for everything.

AK
